doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,441 @@
+| 0x19 | retire_connection_id | | | &check; | &check; | | | | |
+| 0x1A | path_challenge | | | &check; | &check; | | | &check; | |
+| 0x1B | path_response | | | | &check; | | | &check; | |
+| 0x1C | connection_close 0x1C | &check; | &check; | &check; | &check; | &check;
+| 0x1D | connection_close 0x1D | | | &check; | &check; | &check; | | | |
+| 0x1E | handshake_done | | | | &check; | | | | |
+
+The various fields are as defined in RFC 9000.
+
+##### Pkts
+
+_Pkts_ are defined as:
+
+| Pkts | Description|
+| :---: | --- |
+| I | Valid in Initial packets|
+| H | Valid in Handshake packets|
+| 0 | Valid in 0-RTT packets|
+| 1 | Valid in 1-RTT packets|
+
+##### Spec
+
+_Spec_ is defined as:
+
+| Spec | Description |
+| :---: | --- |
+| N | Not ack-eliciting. |
+| C | does not count toward bytes in flight for congestion control purposes. |
+| P | Can be used to probe new network paths during connection migration. |
+| F | The contents of frames with this marking are flow controlled. |
+
+For `C`, `N` and `P`, the entire packet must consist of only frames with the
+marking for the packet to qualify for it.  For example, a packet with an ACK
+frame and a _stream_ frame would qualify for neither the `C` or `N` markings.
+
+### Packets
+
+Frames are coalesced into packets which are then sent by the record layer.
+The `packet_header` is a pointer to the leading bytes of the packet.
+The `frames` are pointers to the individual frames that make up the
+packet's body.
+It is expected that the record layer will encrypt from the `packet_header` and
+`frames` directly without a copy.
+
+```c
+enum packet_validity_e {
+    QUIC_PACKET_INITIAL,
+    QUIC_PACKET_HANDSHAKE,
+    QUIC_PACKET_0_RTT,
+    QUIC_PACKET_1_RTT
+};
+
+typedef enum packet_validity_e PACKET_VALIDITY;
+
+struct ossl_quic_packet_st {
+    QUIC_CONNECTION *conn;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_TXP_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /*
+     * One of the QUIC_PN_SPACE_* values. This qualifies the pkt_num field
+     * into a packet number space.
+     */
+    unsigned int pkt_space : 2;
+
+    /* Pkts options */
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` structure from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- 0-RTT packets can morph into 1-RTT packets and this needs to be handled by
+  the packetiser.
+- the `packet_header` will become a structure / union most likely.
```

Comment:
Some of these notes might need some updates following the latest changes to the above structure.